devtools/gradle/src/main/java/io/quarkus/gradle/workspace/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,239 @@
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ModuleDependency;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.ResolvedConfiguration;
+import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.attributes.Category;
+import org.gradle.api.internal.artifacts.DefaultModuleIdentifier;
+import org.gradle.api.internal.artifacts.dependencies.DefaultExternalModuleDependency;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.tooling.provider.model.ToolingModelBuilder;
+
+import io.quarkus.bootstrap.BootstrapConstants;
+import io.quarkus.gradle.tasks.QuarkusGradleUtils;
+import io.quarkus.gradle.workspace.model.Dependency;
+import io.quarkus.gradle.workspace.model.QuarkusModel;
+import io.quarkus.gradle.workspace.model.impl.DependencyImpl;
+import io.quarkus.gradle.workspace.model.impl.QuarkusModelImpl;
+import io.quarkus.gradle.workspace.model.impl.SourceSetImpl;
+import io.quarkus.runtime.LaunchMode;
+
+public class QuarkusModelBuilder implements ToolingModelBuilder {
+
+    private static final List<String> scannedConfigurations = new LinkedList();
+
+    @Override
+    public boolean canBuild(String modelName) {
+        return modelName.equals(QuarkusModel.class.getName());
+    }
+
+    @Override
+    public Object buildAll(String modelName, Project project) {
+        LaunchMode mode;
+        if (project.getExtensions().getExtraProperties().has("launchMode")) {
+            mode = LaunchMode.valueOf(project.getExtensions().getExtraProperties().get("launchMode").toString());
+        } else {
+            mode = LaunchMode.NORMAL;
+        }
+
+        if (LaunchMode.TEST.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        } else {
+            scannedConfigurations.add(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        }
+
+        if (LaunchMode.DEVELOPMENT.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        }
+
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+        final SourceSet testSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.TEST_SOURCE_SET_NAME);
+
+        final Collection<org.gradle.api.artifacts.Dependency> platformExtension = getPlatformExtension(project);
```

Comment:
I am not sure what platformExtension signifies here. These are application direct extension dependencies, maybe let's call it that?

Reply:
Yes, I agree with you